File: app/src/main/java/com/example/mhmdreza_j/xproject/views/main_page/MainActivity.java
package com.example.mhmdreza_j.xproject.views.main_page;

import android.content.Context;
import android.content.SharedPreferences;
import android.media.MediaPlayer;
import android.os.Bundle;
import android.os.Handler;
import android.provider.Settings;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.app.AppCompatActivity;

import com.example.mhmdreza_j.xproject.R;
import com.example.mhmdreza_j.xproject.application.ApplicationLoader;
import com.example.mhmdreza_j.xproject.utils.SharedPrefUtils;
import com.example.mhmdreza_j.xproject.views.LoadingDialog;
import com.example.mhmdreza_j.xproject.views.base_class.BaseFragment;
import com.example.mhmdreza_j.xproject.views.login.LoginFragment;

import uk.co.chrisjenx.calligraphy.CalligraphyConfig;
import uk.co.chrisjenx.calligraphy.CalligraphyContextWrapper;

import static com.example.mhmdreza_j.xproject.views.login.LoginFragment.IS_USER_LOGGED_IN;
import static java.lang.Thread.sleep;

public class MainActivity extends AppCompatActivity {

    private LoadingDialog loadingDialog;
    private BaseFragment fragment;
    private MediaPlayer player;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        CalligraphyConfig.initDefault(new CalligraphyConfig.Builder()
                .setDefaultFontPath("fonts/LucidaGrandeBold.ttf")
                .setFontAttrId(R.attr.fontPath)
                .build()
        );

        setContentView(R.layout.activity_main);

        if (SharedPrefUtils.getBoolean(IS_USER_LOGGED_IN, false)) {
            startFragment(new MainFragment());
        } else {
            startFragment(new LoginFragment());
        }
        loadingDialog = new LoadingDialog(this);
    }

    @Override
    protected void onResume() {
        super.onResume();
        player = MediaPlayer.create(this, R.raw.background_music);
        player.start();
        player.setLooping(true);
    }

    @Override
    protected void attachBaseContext(Context newBase) {
        super.attachBaseContext(CalligraphyContextWrapper.wrap(newBase));
    }

    public void stopMusic(){
        player.stop();
    }

    public void restartMusic(){
        player.seekTo(0);
    }

    public void startFragment(BaseFragment fragment) {
        this.fragment = fragment;
        FragmentManager manager = getSupportFragmentManager();
        FragmentTransaction fragmentTransaction = manager.beginTransaction();
        fragmentTransaction.replace(R.id.mainFragment, fragment);
        fragmentTransaction.commit();
    }

    @Override
    protected void onStop() {
        super.onStop();
        player.release();
    }

    @Override
    public void onBackPressed() {
        fragment.onBackPressed();
    }

    public void showLoading() {
        loadingDialog.show();
    }

    public void hideLoading() {
        loadingDialog.hide();
    }
}

File: app/src/main/java/com/example/mhmdreza_j/xproject/views/main_page/MainActivity.kt
package com.example.mhmdreza_j.xproject.views.main_page

import android.content.Context
import android.media.MediaPlayer
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import com.example.mhmdreza_j.xproject.R
import com.example.mhmdreza_j.xproject.utils.CANCEL_MATCH
import com.example.mhmdreza_j.xproject.utils.IS_USER_LOGGED_IN
import com.example.mhmdreza_j.xproject.utils.SharedPrefUtils
import com.example.mhmdreza_j.xproject.views.LoadingDialog
import com.example.mhmdreza_j.xproject.views.base_class.BaseFragment
import com.example.mhmdreza_j.xproject.views.login.LoginFragment
import com.example.mhmdreza_j.xproject.webservice.base.constants.BASE_URL
import com.example.mhmdreza_j.xproject.webservice.pref.WebservicePrefSetting
import io.socket.client.IO
import io.socket.client.Socket
import org.json.JSONObject
import uk.co.chrisjenx.calligraphy.CalligraphyConfig
import uk.co.chrisjenx.calligraphy.CalligraphyContextWrapper
import java.net.URISyntaxException

class MainActivity : AppCompatActivity() {

    private var loadingDialog: LoadingDialog? = null
    private var fragment: BaseFragment? = null
    private var player: MediaPlayer? = null
    var socket: Socket? = null


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        CalligraphyConfig.initDefault(CalligraphyConfig.Builder()
                .setDefaultFontPath("fonts/LucidaGrandeBold.ttf")
                .setFontAttrId(R.attr.fontPath)
                .build()
        )

        setContentView(R.layout.activity_main)

        if (SharedPrefUtils.getBoolean(IS_USER_LOGGED_IN, false)) {
            startFragment(MainFragment())
        } else {
            startFragment(LoginFragment())
        }
        loadingDialog = LoadingDialog(this)
    }

    fun startSocket(): Socket? {
        if (socket?.connected() != true) {
            try {
                val opt = IO.Options()
                if (WebservicePrefSetting.getInstance(this).token.isNotEmpty()) {
                    opt.query = "token=${WebservicePrefSetting.getInstance(this).token}"
                }
                socket = IO.socket(BASE_URL, opt)
            } catch (e: URISyntaxException) {
                return null
            }
        }
        socket!!.connect()
        return socket
    }

    fun cancelMatch() {
        socket?.emit(CANCEL_MATCH, JSONObject())
        runOnUiThread { showLoading() }
    }

    override fun onResume() {
        super.onResume()
        startBackgroundMusic()
    }

    override fun attachBaseContext(newBase: Context) {
        super.attachBaseContext(CalligraphyContextWrapper.wrap(newBase))
    }

    private fun stopMusic() {
        if (player != null)
            player!!.stop()
    }

    fun startBackgroundMusic() {
        stopMusic()
        player = MediaPlayer.create(this, R.raw.background_music)
        player!!.start()
        player!!.isLooping = true
    }

    fun startFragment(fragment: BaseFragment) {
        this.fragment = fragment
        val manager = supportFragmentManager
        val fragmentTransaction = manager.beginTransaction()
        fragmentTransaction.replace(R.id.mainFragment, fragment)
        fragmentTransaction.commit()
    }

    fun startWheelMusic() {
        stopMusic()
        player = MediaPlayer.create(this, R.raw.wheel_of_fortune)
        player!!.start()
        player!!.setOnCompletionListener { startBackgroundMusic() }
        player!!.isLooping = false
    }

    fun startGameMusic() {
        stopMusic()
        player = MediaPlayer.create(this, R.raw.game)
        player!!.start()
        player!!.isLooping = true
    }

    override fun onStop() {
        player?.release()
        super.onStop()
    }

    override fun onBackPressed() {
        fragment!!.onBackPressed()
    }

    fun showLoading() {
        loadingDialog!!.show()
    }

    fun hideLoading() {
        loadingDialog!!.hide()
    }
}
